File: 转dat格式/data_define-v5.py
import os
import pandas as pd
import re
from collections import defaultdict

# 初始化设置
pd.set_option('display.max_columns', None)
os.chdir(r"D:\DP小工具\2.转dat格式")

filename = "data"

# ================= 读取CSV文件 =================
try:
    raw = pd.read_csv(
        f"./{filename}.csv",
        dtype=str,
        keep_default_na=False,
        encoding='gbk'
    )
except UnicodeDecodeError:
    raw = pd.read_csv(
        f"./{filename}.csv",
        dtype=str,
        keep_default_na=False,
        encoding='gb18030'
    )


# ================= 列名处理 =================
def sanitize_colname(col):
    return re.sub(r'[._]+', '0', col)


raw.columns = [sanitize_colname(col) for col in raw.columns]


# ================= 判断字段类型 =================
def detect_column_type(series):
    cleaned = series.str.strip()

    # 空值处理
    if cleaned.empty:
        return 'character'

    # 整数检测
    if cleaned.str.match(r'^[+-]?\d+$').all():
        return 'integer'

    # 数值检测（含科学计数法）
    try:
        temp = cleaned.str.replace(',', '', regex=False)
        pd.to_numeric(temp, errors='raise')
        # 排除纯整数
        if temp.str.contains(r'\.|e|E', regex=True).any():
            return 'numeric'
        return 'integer'
    except:
        pass

    return 'character'


# 属性表
attribute = pd.DataFrame({
    'colname': raw.columns,
    'type': [detect_column_type(raw[col]) for col in raw.columns],
    'len': [6] * len(raw.columns)  # 初始化最小长度
})

# ================= 列宽计算 =================
for i, col in enumerate(raw.columns):
    max_len = raw[col].apply(lambda x: len(str(x).strip())).max()
    attribute.at[i, 'len'] = max(6, max_len)


# ================= 数据对齐 =================
def pad_column(series, width):
    return series.str.strip().apply(
        lambda x: x.rjust(width)
    )


for col in raw.columns:
    col_len = attribute[attribute['colname'] == col]['len'].values[0]
    raw[col] = pad_column(raw[col].astype(str), col_len)

# ================= 生成define.stp =================
define_lines = []
cum_pos = 1  # 当前字段起始位置

for _, row in attribute.iterrows():
    start = cum_pos
    end = cum_pos + row['len'] - 1
    # 类型映射
    prefix = {
        'integer': 'di',
        'numeric': 'dw',
        'character': 'dc'
    }[row['type']]
    define_lines.append(
        f"{prefix} ${row['colname']}=${start}-{end},"
    )
    cum_pos = end + 2  # 字段宽度 + 1个空格分隔符

# ================= 生成make.stp =================
# 多选题识别逻辑
pattern = re.compile(r'^(.*?)(0\d+)+$')
base_counts = defaultdict(int)

for col in raw.columns:
    match = pattern.match(col)
    if match:
        base_name = match.group(1)
        base_counts[base_name] += 1

# 过滤有效多选字段（出现次数>1）
multi_choice = {k: v for k, v in base_counts.items() if v > 1}

# 构建make.stp内容
make_lines = ["[*data ttl(;)=="]
for name, count in multi_choice.items():
    make_lines.append(f"{name};{count};")
make_lines += [
    "]",
    "[*do i=1:[ttl.#]/2]",
    "   [*do a=1:[ttl.i*2]]",
    "      om $[ttl.i*2-1]=$[ttl.i*2-1]0[a]/1-999,",
    "   [*end a]",
    "[*end i]"
]

# ================= 文件输出 =================
# 输出dat文件
with open(f"{filename}1.dat", 'w', encoding='gbk') as f:
    for _, row in raw.iterrows():
        line = ' '.join(row.values)  # 字段间添加空格
        f.write(line + '\n')

# 输出配置文件
with open(f"{filename}1define.stp", 'w', encoding='gbk') as f:
    f.write('\n'.join(define_lines))

with open(f"{filename}1make.stp", 'w', encoding='gbk') as f:
    if multi_choice:
        f.write('\n'.join(make_lines))
    else:
        f.write("No multiple choice fields detected")

print("处理完成！字符型列信息：")
print(attribute['type'].value_counts())
print(attribute[attribute['type'] == 'character'])

File: 转dat格式/data_define-v5.R
rm(list=ls())
#options(digits = 22)
setwd("D:\\DP小工具\\2.转dat格式")
filename<-c("data")				#文件名字
raw<-read.csv(paste("./",filename,".csv",sep=""),stringsAsFactors = F)
raw[,which(lapply(raw,class)=='numeric')] <- round(raw[,which(lapply(raw,class)=='numeric')],5)
names(raw) <- gsub("[.]","0", names(raw))#将"."全部替换成0
attribute<-data.frame(colname=colnames(raw),type=0)
attribute[,1]<-gsub("_",0,attribute[,1])

for(i in 1:ncol(raw)){
  if(class(raw[,i])=="character"){
    attribute[i,2]<-"character"
  }
  if(class(raw[,i])=="numeric"){
    attribute[i,2]<-"numeric"
  }
  if(class(raw[,i]) %in% c("character","numeric")==F){
    attribute[i,2]<-"integer"
  }
}
raw[is.na(raw)]<-" "
##转dat格式
len <- c()
for(i in 1:ncol(raw)){
  len[i] <- max(nchar(raw[,i],type = "bytes"))
}
len[len<=6]<-6
attribute<-cbind(attribute,len,row.names =1:length(len))
#不直接用上面计算的,重新计算一次,因为两个for循环太慢,自定义一个函数快很多
fix_data<-function(df){
  df1<-c()
  len1<-max(nchar(df,type = "bytes"))
  len1[len1<=6]<-6
  for(i in 1:length(df)){
    df1[i]<-paste(paste(rep(" ",len1-nchar(df[i],type = "bytes")),collapse=""),df[i],sep="")
  }
  return(df1)
}
#用空格填充实现等宽数据
for(i in 1:ncol(raw)){
  raw[,i]<-fix_data(raw[,i])
}

##define.stp
s<-c()
for(i in 1:nrow(attribute)){
  s[i]<-(cumsum(attribute[1:i,3])+i)[i-1]
}
s[1]<-1
e<-s+attribute[,3]-1
attribute$define<-0
for(i in 1:nrow(attribute)){
  if(attribute[i,2]=="integer"){
    attribute[i,4]<-paste("di $",attribute[i,1],"=$",s[i],"-",e[i],",",sep="")
  }
  if(attribute[i,2]=="numeric"){
    attribute[i,4]<-paste("dw $",attribute[i,1],"=$",s[i],"-",e[i],",",sep="")
  }
  if(attribute[i,2]=="character"){
    attribute[i,4]<-paste("dc $",attribute[i,1],"=$",s[i],"-",e[i],",",sep="")
  }
}
##make.stp
colname1<-colnames(raw)
colname2<-gsub("_.*","",colname1)
colname2<-factor(colname2,unique(colname2))
colname_num<-as.data.frame(table(colname2))
m_colname<-colname_num[colname_num$Freq>1,]
make<-data.frame(df=paste(m_colname[,1],";",m_colname[,2],";",sep=""),stringsAsFactors = F)
if(make[1,1]==";;"){
  make<-"No multiple choice,please check your input data."
}else{
  make<-rbind(c("[*data ttl(;)="),make,"]","[*do i=1:[ttl.#]/2]","   [*do a=1:[ttl.i*2]]","      om $[ttl.i*2-1]=$[ttl.i*2-1]0[a]/1-999,","   [*end a]","[*end i]",sep="")
}
##输出结果
write.table(raw,paste(filename,".dat",sep=""),row.names = F,col.names = F,quote = F)
write.table(attribute[,4],paste(filename,"define.stp",sep=""),row.names = F,col.names = F,quote = F)
write.table(make,paste(filename,"make.stp",sep=""),row.names = F,col.names = F,quote = F)
attribute[attribute$type=="character",]#查看文本型字段


